File: frontend/src/types/budget.ts
// ── Budget response types matching backend BudgetStatusDto / BudgetOverviewDto ──

/** One category's budget status with actual spending (from /api/budget/overview) */
export interface Budget {
  categoryId: string;
  categoryName: string;
  categoryIcon?: string;
  categoryColor?: string;

  // budget info (null when no budget set)
  budgetId?: string;
  hasBudget: boolean;
  plannedAmount: number;

  // actual spending from transactions
  spentAmount: number;       // actualAmount / đã chi

  // computed by backend
  remainingAmount: number;
  usedPercent: number;       // percentUsed
  isOverBudget: boolean;
  isNearLimit: boolean;
  isNoBudget: boolean;
}

/** Full budget CRUD record (from /api/budget GET) */
export interface BudgetRecord {
  id: string;
  categoryId: string;
  categoryName: string;
  year: number;
  month: number;
  plannedAmount: number;
  note?: string;
}

export interface CreateBudgetDto {
  categoryId: string;
  year: number;
  month: number;
  plannedAmount: number;
  note?: string;
}

export interface UpdateBudgetDto {
  plannedAmount?: number;
  note?: string;
}

/** Overview for a single month (from /api/budget/overview) */
export interface BudgetOverview {
  year: number;
  month: number;
  monthName: string;

  // totals
  totalPlanned: number;
  totalSpent: number;        // backend: totalSpent
  totalRemaining: number;
  usedPercent: number;

  // stats
  overBudgetCount: number;
  nearLimitCount: number;
  onTrackCount: number;
  noBudgetCount: number;

  // per-category breakdown
  categories: Budget[];
}

export interface BudgetYearlyOverview {
  year: number;
  months: BudgetMonthSummary[];
}

export interface BudgetMonthSummary {
  month: number;
  totalPlanned: number;
  totalSpent: number;
  totalRemaining: number;
  usedPercent: number;
  isOverBudget: boolean;
  budgetCount: number;
  overBudgetCount: number;
}

export interface BulkUpsertBudgetItemDto {
  categoryId: string;
  plannedAmount: number;
}

export interface BulkUpsertBudgetDto {
  year: number;
  month: number;
  items: BulkUpsertBudgetItemDto[];
}

export interface CopyBudgetDto {
  fromYear: number;
  fromMonth: number;
  toYear: number;
  toMonth: number;
  overwrite: boolean;
}

export interface CopyBudgetResult {
  message: string;
  created: number;
  skipped: number;
  overwritten: number;
}

File: ios/ExpenseTrackerApp/Shared/Models/Budget.swift
import Foundation

// MARK: - Budget CRUD Record (from GET /api/budget)

struct BudgetRecord: Codable, Identifiable {
    let id: String
    let categoryId: String
    let categoryName: String
    let year: Int
    let month: Int
    let plannedAmount: Double
    let note: String?
}

// MARK: - Budget Status (per-category item in overview)

struct BudgetStatus: Codable, Identifiable {
    var id: String { categoryId }
    let categoryId: String
    let categoryName: String
    let categoryIcon: String?
    let categoryColor: String?
    let budgetId: String?
    let hasBudget: Bool
    let plannedAmount: Double
    let spentAmount: Double
    let remainingAmount: Double
    let usedPercent: Double
    let isOverBudget: Bool
    let isNearLimit: Bool
    let isNoBudget: Bool
}

// MARK: - Monthly Budget Overview (from GET /api/budget/overview)

struct BudgetOverview: Codable {
    let year: Int
    let month: Int
    let monthName: String?
    let totalPlanned: Double
    let totalSpent: Double
    let totalRemaining: Double
    let usedPercent: Double
    let overBudgetCount: Int
    let nearLimitCount: Int
    let onTrackCount: Int
    let noBudgetCount: Int
    let categories: [BudgetStatus]
}

// MARK: - Yearly Budget Summary (from GET /api/budget/overview/yearly)

struct MonthlyBudgetSummary: Codable, Identifiable {
    var id: Int { month }
    let month: Int
    let totalPlanned: Double
    let totalSpent: Double
    let totalRemaining: Double
    let usedPercent: Double
}

// MARK: - Request Models

struct CreateBudgetRequest: Codable {
    let categoryId: String
    let year: Int
    let month: Int
    let plannedAmount: Double
    let note: String?
}

struct UpdateBudgetRequest: Codable {
    let plannedAmount: Double
    let note: String?
}

struct CopyBudgetRequest: Codable {
    let sourceYear: Int
    let sourceMonth: Int
    let targetYear: Int
    let targetMonth: Int
    let overwrite: Bool
}

struct CopyBudgetResponse: Codable {
    let message: String?
    let created: Int
    let skipped: Int
    let overwritten: Int
}
